### DISCUSSION TITLE
Alt-Tab behaviour

### DISCUSSION THREAD

**Discussion by sandoru** [cajhin/capsicain#74] [Q&A]
Is there some way to implement Alt-Tab window switching behaviour with a virtual modificator (e. g. MOD9) instead of Alt?

**Comment by cajhin** [comment]
The Alt-Tab combo is tricky, because (from what Windows sees) Alt must stay down, and no other key than Tab must be pressed.
With a normal keyCombo() you send a single Alt-v Tab-v^ Alt-^, which does not work.

The only way that I know that works is that your modifier key must be the Alt key (either the real unmodified Alt, or another key that is rewired to Alt). Then you can press and hold that, and press either Tab, or a key that is mapped to Tab.

You could, for example, make Ctrl+T act as Alt-Tab:
REWIRE LCTRL LALT 
COMBO T [&...] > key(TAB)

My own config does this:
REWIRE TAB MOD10 TAB               # Tab to Mod10, but send Tab when tapped
REWIRE LALT MOD12  //  LALT     # LAlt to Mod12, but revert to LAlt when tap+hold (press,release, press, hold)
This is quite complicated, but it actually works fine.

**Comment by sandoru** [comment]
Thanks! The drawback is, one can incidentally press and release MOD12 (having changed one's mind on some action) and get quiet Alt waiting instead of MOD12.

Could be maybe something like this implemented?

When receiving MOD12 + Tab capsicain:
    a) sends MOD12_Up to itself;
    b) set some "AltTabMode" flag for itself activating **temporary combo for MOD12_Up event**;
    c) sends LAlt_Down, then Tab_Down to Windows;
    d) after receiving MOD12_Up sends LAlt_Up to Windows (the said **temporary combo**) and clear "AltTabMode" flag.

It can't be done through switching configs now, as up-only events aren't supported.

Another idea: option to translate virtual modifiers to Windows as some unused real keys (like F24). It'd allow to use them in AHK where Alt-Tab simulation function exists.

Is one of these viable?

**Reply by cajhin** [reply]
that would be a lot of special code for one specific combo. What you get is basically "Alt on tap+hold".
REWIRE yourKey MOD12 // LALT

Note the //. That's "on tap, do nothing". 
Alt only triggers when you tap&hold (press-release-press-hold) your key. And if you release that without action, well, it's the same when you press Alt and then decide to do nothing.

**Comment by a13ph** [comment]
# workarounds

It is probably easier to use ahk for handling alt-tab, since it already has some special handling for this combo (at least it had in v1 and vL years ago)


I would suggest foregoing it altogether, since it is a lacklaster way of navigating windows 


imo better options are:
- switching directly to necessary app by win+icon number or external launcher app
- switching without preview with alt+esc or some similar external hotkey


also:
- ahk had lots of scripts as altttab alternatives
- 7 taskbar tweaker might also be of interest for managing how taskbar works,
- 7 taskbar numberer in positional mode can improve your win+num experience
- powertoys run or flow launcher can be decent options if you need window switching, see window walker plugin
- tiling wms and hotkey daemons now exist for ms windows
